### ISSUE TITLE
Nightly rejects some macro-generated doc attribute values accepted by stable

### ISSUE THREAD

**Issue by Erutuon** [rust-lang/rust#85432]
A [version](https://github.com/Erutuon/parse-mediawiki-sql/tree/adfe3364fafaee21fafb22b28ba0fdb44c89c59d) of my crate [parse-mediawiki-sql](https://crates.io/crates/parse-mediawiki-sql) with macro-generated docs builds on stable, but errors out on nightly, where the error suggests that I fix the error by enabling `#![feature(extended_key_value_attributes)]`.

Stable is cargo 1.52.0 (rust-lang/cargo@69767412a 2021-04-21), rustc 1.52.1 (9bc8c42bb 2021-05-09), nightly is cargo 1.54.0-nightly (rust-lang/cargo@070e459c2 2021-05-11), rustc 1.54.0-nightly (3e99439f4 2021-05-17).

The error occurs in a macro that generates a struct declaration with an implementation of a trait and calls another macro to take a string generated by another set of macros and put it into a doc attribute above the struct. [Here](https://github.com/Erutuon/parse-mediawiki-sql/blob/adfe3364fafaee21fafb22b28ba0fdb44c89c59d/src/schemas.rs#L31-L200) are the lines containing the macro declarations.

The error looks like this:

```
error[E0658]: arbitrary expressions in key-value attributes are unstable
   --> src/schemas.rs:152:13
    |
152 |               database_table_doc!($table_name $(, $page)?),
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
274 | / impl_row_from_sql! {
275 | |     image
276 | |     Image<'input> {
277 | |         name: PageTitle,
...   |
293 | |     }
294 | | }
    | |_- in this macro invocation
    |
    = note: see issue #78835 <https://github.com/rust-lang/rust/issues/78835> for more information
    = help: add `#![feature(extended_key_value_attributes)]` to the crate attributes to enable
    = note: this error originates in the macro `impl_row_from_sql` (in Nightly builds, run with -Z macro-backtrace for more info)
```

There are 22 calls to the macro, but only 13 errors, for the calls those whose contents start with `$table_name:ident $output_type:ident<$life:lifetime>`. I tried to generate a minimal example by copying the macros and one of the failing calls, but failed: [this reduced version](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4c186a7017530eeb6c506f3417e9bcf6) compiled on both stable and nightly in the playground.

Somehow this sequence of tokens causes the error in the crate but not in the smaller playground example. I don't know much about the internals of the compiler here so have no idea how to change the playground example to generate the error.

**Comment by hellow554** [comment]
While compiling with nightly another (maybe (un-)related error) pops up, which isn't present on stable:

```
[...]
error[E0308]: mismatched types
    --> /home/marcel/.cargo/registry/src/github.com-1ecc6299db9ec823/lexical-core-0.7.4/src/atof/algorithm/math.rs:2065:27
     |
2065 |     let rs = Limb::BITS - s;
     |                           ^ expected `u32`, found `usize`

error[E0277]: cannot subtract `usize` from `u32`
    --> /home/marcel/.cargo/registry/src/github.com-1ecc6299db9ec823/lexical-core-0.7.4/src/atof/algorithm/math.rs:2065:25
     |
2065 |     let rs = Limb::BITS - s;
     |                         ^ no implementation for `u32 - usize`
     |
     = help: the trait `Sub<usize>` is not implemented for `u32`
```

**Comment by hellow554** [comment]
Here's what I got:

```rust
macro_rules! with_doc_comment {
    ($comment:expr, $item:item) => {
        #[doc = $comment]
        $item
    };
}

macro_rules! database_table_doc {
    () => {
        ""
    };
}

with_doc_comment! {
    database_table_doc!(),
    #[derive(Debug)]
    struct Image {
        #[cfg(test)]
        _f: (),
    }

}

fn main() {}
```

I had a similar piece of code (I try to post it in the next post) that fails with the exact message on stable as well.

But here's the regression:

searched nightlies: from nightly-2019-10-01 to nightly-2021-05-01
regressed nightly: nightly-2021-04-12
searched commits: from https://github.com/rust-lang/rust/commit/a836d9b6413d9d593be6c09463ff8c4c70e56599 to https://github.com/rust-lang/rust/commit/a8661245649f3d1c0dc5b23270bdac0bbd2d8f64
regressed commit: https://github.com/rust-lang/rust/commit/ba6275b6a9fc05fd5d93220e9a67fe64d663cb62

cc #82608 @Aaron1011

**Comment by hellow554** [comment]
This code fails **on stable** with the same message as on nightly:

```rust
macro_rules! database_table_doc {
    () => {
        ""
    };
}

#[doc = database_table_doc!()]
#[derive(Debug)]
struct Image {
    #[cfg(test)]
    _f: (),
}

fn main() {}
```

I just "inlined" the `with_doc_comment` macro.

On an older version (2019-10-01) it fails with:

```
error: unexpected token: `database_table_doc`
 --> src/main.rs:7:9
  |
7 | #[doc = database_table_doc!()]
  |         ^^^^^^^^^^^^^^^^^^
```

Going to try a specific comment right now...

**Comment by hellow554** [comment]
Funny enough, in the [tracking issue #78835](https://github.com/rust-lang/rust/issues/78835#issuecomment-797994243) @alexschrod had the same issue as in this post, so this error is somewhat known. In the paste crate it was reported at https://github.com/dtolnay/paste/issues/29 and fixed in https://github.com/dtolnay/paste/pull/30, so maybe that's something for you @Erutuon if you need a fix right now.

**Comment by hameerabbasi** [comment]
Sorry for the spam!

@rustbot modify labels -ICEBreaker-Cleanup-Crew

**Comment by hellow554** [comment]
@hameerabbasi you can edit your [post](https://github.com/rust-lang/rust/issues/85432#issuecomment-843235352) and use the rustbot command again without creating a new post ;)

**Comment by Erutuon** [comment]
> This code fails **on stable** with the same message as on nightly:
> 
> ```rust
> macro_rules! database_table_doc {
>     () => {
>         ""
>     };
> }
> 
> #[doc = database_table_doc!()]
> #[derive(Debug)]
> struct Image {
>     #[cfg(test)]
>     _f: (),
> }
> 
> fn main() {}
> ```
> 
> I just "inlined" the `with_doc_comment` macro.

Right, that's the reason for the [`with_doc_comment` macro](https://github.com/Erutuon/parse-mediawiki-sql/blob/adfe3364fafaee21fafb22b28ba0fdb44c89c59d/src/schemas.rs#L45-L53). On stable, taking a macro-generated string followed by an item and passing them another macro and having that macro add the string to a doc attribute above the item seems to convince the compiler (somehow) that the doc attribute value is acceptable, even though it's not really a string literal.

> Funny enough, in the [tracking issue #78835](https://github.com/rust-lang/rust/issues/78835#issuecomment-797994243) @alexschrod had the same issue as in this post, so this error is somewhat known. In the paste crate it was reported at [dtolnay/paste#29](https://github.com/dtolnay/paste/issues/29) and fixed in [dtolnay/paste#30](https://github.com/dtolnay/paste/pull/30), so maybe that's something for you @Erutuon if you need a fix right now.

It looks like I'm using the same technique as the [regression test](https://github.com/dtolnay/paste/blob/dd8fc9ff65aa228ffae3e5e084629adc76c262b5/tests/test.rs#L309-L327), but on nightly now it doesn't compile in 13 out of the 22 macro calls in my code.

**Comment by Aaron1011** [comment]
This will be fixed by https://github.com/rust-lang/rust/pull/85445

**Comment by hameerabbasi** [comment]
Assigning `P-critical` as discussed in the [WG-prioritization discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2385432.20Nightly.20rejects.20some.20macro-generated.20doc.20attribute.20v.E2.80.A6) and removing I-prioritize.

@rustbot modify labels +P-critical -I-prioritize

**Comment by pietroalbini** [comment]
Fixed on 1.54 by https://github.com/rust-lang/rust/pull/85445 and fixed on 1.53 by https://github.com/rust-lang/rust/pull/86225.
